client/src/util/centroid.js

```diff
@@ -1,61 +1,125 @@
+    const categoryValue = categoryArray[i];
+
+    // Get the index of the categoryValue within the category
+    // If the category is truncated and this value is removed,
+    // it will not be assigned a category value and will not be
+    // labeled on the graph
+    const categoryValueIndex = categoricalSelection[
+      categoryName
+    ].categoryValueIndices.get(categoryValue);
 
-  const annoArray = world.obsAnnotations.col(annoName).asArray();
-  const layoutXArray = world.obsLayout.col(layoutDimNames[0]).asArray();
-  const layoutYArray = world.obsLayout.col(layoutDimNames[1]).asArray();
+    // Check to see if the current category is a user created annotation
+    // if the user created this category, do not create a label for the `unassigned` value
+    const isUserAnno = schemaObsByName[categoryName].writable;
+
+    if (
+      categoryValueIndex !== undefined &&
+      !(isUserAnno && categoryValue === unassignedCategoryLabel)
+    ) {
+      // Get the number of cells which are in the category value
+      const numInCategoryValue =
+        categoricalSelection[categoryName].categoryValueCounts[
+          categoryValueIndex
+        ];
+
+      // Create/fetch the valueArray,
+      // which is what the key points to in the `coordinates` hashmap
+      const valueArray = coordinates.get(categoryValue) || [
+        false, // hasFinte
+        0, // index
+        new Float32Array(numInCategoryValue), // x coorindates
+        new Float32Array(numInCategoryValue) // y coorindates
+      ];
+      const index = valueArray[1];
+      let hasFinite = valueArray[0];
 
-  for (let i = 0, len = annoArray.length; i < len; i += 1) {
-    if (annoArray[i] === annoValue) {
       hasFinite =
-        Number.isFinite(layoutXArray[i]) || Number.isFinite(layoutYArray[i])
+        Number.isFinite(layoutXArray[i]) && Number.isFinite(layoutYArray[i])
           ? true
           : hasFinite;
-      centroidX.push(layoutXArray[i]);
-      centroidY.push(layoutYArray[i]);
+      valueArray[0] = hasFinite;
+      valueArray[1] = index + 1;
+      valueArray[2][index] = layoutXArray[i];
+      valueArray[3][index] = layoutYArray[i];
+
+      coordinates.set(categoryValue, valueArray);
     }
   }
 
-  if (hasFinite) {
-    const medianX = quantile([0.5], Float64Array.from(centroidX));
-    const medianY = quantile([0.5], Float64Array.from(centroidY));
-
-    return [medianX, medianY];
+  // Iterate over the recently created map
+  const iter = coordinates.entries();
+  let pair = iter.next().value;
+  let value;
+  let key;
+  while (pair) {
+    key = pair[0];
+    value = pair[1];
+    // If there are coordinates for this cateogrical value,
+    // and there is a finite coordinate for the category value
+    if (value[2].length > 0 && value[3].length > 0 && value[0]) {
+      // Find the median x and y coordinate
+      // and insert them into the first two indices
+      value[0] = quantile([0.5], value[2])[0];
```

Comment:
have you confirmed that the use of median (vs. mean) is worth the significant computational overhead?

Reply:
Define significant? :)

It does make a really big difference in keeping the labels over the center of mass of the cluster. Mean is strongly influence by outliers, which tends to strongly pull the labels towards the "stray" cells and off the main cluster itself.

Reply:
Significant - in UI delay/lag due to median calc cost --  (very roughly) takes this moves us from an O(N) to O(NlogN) [best case] overhead.  Where cell count is high, we will see UI lag.  Just want to confirm that the (still suboptimal) median is worth the cost.   Fully good with the answer being "yes, worth it" :-)